### ISSUE TITLE
Kernel freeze when KOReader going to sleep with WLAN enabled

### ISSUE THREAD

**Issue by S-trace** [koreader/koreader#3626]
* KOReader version: koreader-nightly-20180110
* Device: Kobo Aura H2O, FW 4.6.10075

#### Issue
Device freezes after entering mem sleep when running Koreader with WLAN enabled

#### Steps to reproduce
1. Start KOReader
2. Make sure that WLAN is enabled
3. Press Power button
4. Wait ~18 seconds
5. Try to wake up device - it is now frozen (neither power button, sleep cover or USB cable plug/unplug can't wake it up - only long power button press or reset button can reboot device)

##### `crash.log` (if applicable)
[crash.log](https://github.com/koreader/koreader/files/1650645/crash.log)

Log in UART console (with no_console_suspend added to kernel cmdline in U-Boot environment) says:
```
[   97.740036] PM: Preparing system for mem sleep
</>[   97.744559] Freezing user space processes ... </>[   97.762460] (elapsed 0.01 seconds) </>[   97.766041] Freezing remain
ing freezable tasks ... (elapsed 0.01 seconds) done.[   97.784903] PM: Entering mem sleep
[   97.832460] Waiting for arm_podf changed!
[   97.862477] Waiting for arm_podf changed!
[   97.992647] USB Host suspend begins
[   97.996182] ehci_fsl_drv_suspend, pm event 
[   98.001685] WARNING: driver bcmsdh_sdmmc did not remove its interrupt handler!
[   98.009698] Unable to handle kernel NULL pointer dereference at virtual address 00000004
[   98.017851] pgd = 9e574000
[   98.020590] [00000004] *pgd=8e53f031, *pte=00000000, *ppte=00000000
[   98.026978] Internal error: Oops: 817 [#1] PREEMPT
[   98.031805] last sysfs file: /sys/power/state
</>[   98.036198] Modules linked in: dhd sdio_wifi_pwr [last unloaded: arcotg_udc][   98.044064] Process reader.lua (pid: 951,
 stack limit = 0x9ec2a2e8)
[   98.050382] Stack: (0x9ec2bd40 to 0x9ec2c000)
[   98.054808] bd40: 9e147800 8021af58 9ec2a000 00000000 9e21ae60 80053474 00000000 00000000
[   98.063064] bd60: 9e275580 7f03ef28 7f03ef30 00000000 9e0b2c44 00000000 1dbf2d40 7f0373bc
[   98.071320] bd80: 9ec4d800 9e706180 7f03ef30 7f037e44 9ec4d800 7f0320d4 7f03ef1c 9e706000
[   98.079577] bda0: 7f03ef30 7f037fec 9e931000 7f03cfa0 803c3f70 80220108 9e931008 7f03cfa0
[   98.087832] bdc0: 803c3f70 801bead0 9e93103c 9e931008 803c3f70 801bec20 9e931008 00000064
[   98.096089] bde0: 803c3f70 801be14c 9e931008 9e147c08 0000006f 801bc910 9e931008 9e1e2000
[   98.104343] be00: 0000006f 80220014 00000001 8021f4dc 9e1e2000 ffffffda 9e0b2c40 00000002
[   98.112600] be20: 9e0b2c44 8021b358 00000000 9e0b2c40 9e0b2c40 80224090 9ec2be60 803976b0
[   98.120854] be40: 802c5ef4 00000002 00000000 801bfd6c 9ec2be60 801c23b0 9ec2be97 00000000
[   98.129112] be60: 1dbf2d40 00000061 9ec2be97 803976b0 00000002 803976e4 803c033c 803df048
[   98.137365] be80: 803a9c94 00000000 00000002 801c2d00 803976b0 00000002 00000000 00000000
[   98.145621] bea0: 803976b0 80397708 803c033c 801c3090 110df21c 00000061 110df21c 00000061
[   98.153877] bec0: 803975f0 9e504268 00000003 803322e4 003fffff 00000000 00000003 802a9dcc
[   98.162131] bee0: 802bd030 803322e4 2b9a41d0 80083374 803322e4 00000003 00000000 00000004
[   98.170401] bf00: 802a9dcc 800835d0 9e536000 00000003 00000003 80082cf8 9e0195d0 9e97d598
[   98.178657] bf20: 9ec2bf80 00000004 9e97d580 802bd030 9e002880 80168374 9e0195d0 80100cc4
[   98.186913] bf40: 00000004 9e26e280 2aac5000 9ec2bf80 00000004 9ec2a000 2b9a4268 800b6b1c
[   98.195167] bf60: 9e26e280 2aac5000 9e26e280 2aac5000 00000000 00000000 00000004 800b6c70
[   98.203420] bf80: 00000000 00000000 00000004 00000000 00000004 2aac5000 000f5e10 00000004
[   98.211674] bfa0: 80026104 80025f80 00000004 2aac5000 00000008 2aac5000 00000004 00000000
[   98.219930] bfc0: 00000004 2aac5000 000f5e10 00000004 2ac731c0 2b9a41c8 2b9a4268 2b9a41d0
[   98.228185] bfe0: 00000000 7ef87464 2abd32fb 2ac0dfe6 400e0030 00000008 00000000 00000000
[   98.239798] Code: e5924004 e2844001 e5824004 e5902000 (e5823004)
[   98.246166] ---[ end trace c3f170c9c4eee21d ]---
[   98.250859] note: reader.lua[951] exited with preempt_count 1
[   98.256834] BUG: scheduling while atomic: reader.lua/951/0x40000002
</>[   98.263200] Modules linked in: dhd sdio_wifi_pwr [last unloaded: arcotg_udc][   98.275583] Waiting for arm_podf changed!
[   98.342468] Waiting for arm_podf changed!
[   98.382483] Waiting for arm_podf changed!
```
Looks like that WLAN kernel modules must be unloaded before suspending device, and then reloaded after resume.

**Comment by Frenzie** [comment]
Is that the H2O v1 or v2?

See disable-wifi.sh which is executed before suspend.

**Comment by S-trace** [comment]
Thank you for reply. This is H2O v1.
I have now added some debug lines like **`echo "<1> disable_wlan.sh removing ${WIFI_MODULE}" > /dev/kmsg`** to disable-wifi.sh, checked that this debug output is printed to UART console.

First test:
When disabling WLAN using KOReader menu i can see messages from disable_wlan.sh when it is started, no kernel freeze happened.

Second test:
If WLAN was disabled and then re-enabled in KOReader menu, disable_wlan.sh is started on suspend, and no kernel freeze happened.

Third test: 
1. Extracted clean copy of koreader/ directory from koreader-kobo-arm-kobo-linux-gnueabihf-v2015.11-1457-g7feea40.zip
2. Replaced disable_wlan.sh by patched for debug logging version
3. Started KOReader from KSM9
4. Dismissed "Conversion complete" message
5. Pressed power button on "KOReader Quickstart Guide" page
Result: no debug messages from disable_wlan.sh was printed, no modules was unloaded and kernel freeze happened when KOReader tried to put device to mem sleep.

**Comment by Frenzie** [comment]
There were some changes in https://github.com/koreader/koreader/pull/2215 and https://github.com/koreader/koreader/pull/2705 but my own H2O has been suspending perfectly fine for the past year. I've definitely never encountered any freezes in the past either, just a failure to suspend.

**Comment by poire-z** [comment]
Totally off-topic, just taking a chance :)
@S-trace : you have a recent koreader running on a recent kobo firmware, and Dictionary lookup should fail with a segmentation fault when running the `sdcv` binary (see #3302). Is that your case?
If yes, and as you seem at ease on linux, is there any usable info in kernel logs when that happen? And if you have tools (catchsegv, and given your username, may be you have `strace` :), any other information you can get that could help us?

**Comment by Frenzie** [comment]
strace is a piece of cake on Kobo. It's not like it's Android or anything. ;-) It's close enough to proper Linux, although I think that Linus might say something like [this](http://lkml.iu.edu/hypermail/linux/kernel/1801.2/04628.html) about some of their touch drivers. But that aside.

Perhaps it could be the libc version?

https://github.com/koreader/koxtoolchain/blob/47acb959b6169b5d1e546814d1747a8f88484b1b/configs/ct-ng-kobo-config#L293

That was explicitly changed from 2.11 to 2.15 in the past as a fix, though I don't know as a fix for what: https://github.com/koreader/koxtoolchain/commit/d640340a96c76f0271b803d03af8bf8dc2fc20dc

My personal computer has 2.24, but of course the differences go much deeper than that.

```
$ apt policy libc6
libc6:
  Installed: 2.24-11+deb9u1
  Candidate: 2.24-11+deb9u1
  Version table:
     2.26.9000+20180108.401311cf-0experimental0 1
          1 http://ftp.be.debian.org/debian experimental/main amd64 Packages
     2.26-4 1
          1 http://ftp.be.debian.org/debian unstable/main amd64 Packages
 *** 2.24-11+deb9u1 500
        500 http://ftp.be.debian.org/debian stretch/main amd64 Packages
        500 http://security.debian.org stretch/updates/main amd64 Packages
        100 /var/lib/dpkg/status
``1

**Comment by S-trace** [comment]
> Perhaps it could be the libc version?
Yes, looks like this. 
Replied to #3302 with some research around this issue

**Comment by S-trace** [comment]
The kernel freeze problem is only present if there is no wifi_was_on parameter in settings.reader.lua
It is added only after toggling WLAN off and back on, and then exit from KOReader - after this, WLAN modules are unloaded.
You can test this by moving your settings.reader.lua somewhere, then starting KOReader with WLAN enabled and putting it to sleep state.

**Comment by Frenzie** [comment]
Hm, not sure what the point of that check is; preventing some harmless log noise perhaps. Since I'm on my phone atm would you be able to give removing the test from 2215 I referenced above a quick try? It's in device/generic iirc.

**Comment by S-trace** [comment]
> would you be able to give removing the test from 2215 I referenced above a quick try?

Yes, removing that check fixed this issue for me

**Comment by Frenzie** [comment]
I meant for debugging though just like with the frontlight it's a simple matter of it should always be disabled no matter what. A freeze or even a simple no suspend in an edge case is much more serious than no reconnect in an edge case after all.

We could check if the kernel module is loaded on startup to set the var maybe? Like how KSM does the wifi is on warning but with the opposite purpose.

**Comment by Frenzie** [comment]
I just checked, KSM doesn't do anything fancy but simply the most basic thing you can think of:

```bash
lsmod | grep -c sdio_wifi_pwr
```

So at its simplest that would be something like (in `NetworkManager:init()`)

```lua
self.wifi_was_on = os.execute("lsmod | grep -c sdio_wifi_pwr") or G_reader_settings:isTrue("wifi_was_on")
```

So I propose to do that or something much like it combined with removing the check prior to entering suspend to do our utmost best to make sure wifi is always off just like we used to.

**Comment by KenMaltby** [comment]
It may just be me, but I don't see a reason to have WiFi on, unless there is a function that needs it.  If you guys someday get to the point where an autostart/end of the Wifi around a function that needs it, is working, the issue will be mute.

**Comment by poire-z** [comment]
Should be fixed by #3939
